lib/core/macros.nim

```diff
@@ -1344,6 +1348,93 @@ proc copy*(node: NimNode): NimNode {.compileTime.} =
   ## An alias for copyNimTree().
   return node.copyNimTree()
 
+type GenAstOpt* = enum
+  kNoExposeLocalInjects,
+    # when unset, inject'd symbols (including implicit ones such as local procs
+    # in scope) are exposed implicitly;
+    # gensym'd symbols will generate a CT internal error: `environment misses:`
+    # when set, local symbols are not exposed unless captured explicitly in
+    # `genAst` argument list. The default is unset, to avoid surprising hijacking
+    # of local symbols by symbols in caller scope.
+
+macro genAst*(options: static set[GenAstOpt] = {}, args: varargs[untyped]): untyped =
+  ## Accepts a list of captured `variables = value` and a block and returns the
+  ## AST that represents it. Local `{.inject.}` symbols are captured (eg
+  ## local procs) unless `kNoExposeLocalInjects in options`; additional variables
+  ## are captured as subsequent parameters.
+  runnableExamples:
+    type Foo = enum kfoo0, kfoo1, kfoo2, kfoo3, kfoo4
+
+    macro bar1(x0: static Foo, x1: Foo, x2: Foo, xignored: Foo): untyped =
+      let s0 = "not captured!" ## does not override `s0` from caller scope
+      let s1 = "not captured!" ## does not override `s1=2`
+      let xignoredLocal = kfoo4
+      proc localExposed(): auto = kfoo4 # implicitly captured
+      let x3 = newLit kfoo4
+      result = genAst({}, s1=2, s2="asdf", x0=newLit x0, x1=x1, x2, x3):
+        # echo xignored # would give: Error: undeclared identifier
+        # echo s0 # would give: Error: internal error: expr: var not init s0_237159
+        (s1, s2, x0, x1, x2, x3, localExposed())
+
+    macro bar2(x0: static Foo, x1: Foo, x2: Foo, xignored: Foo): untyped =
+      let s0 = "not captured!" ## does not override `s0` from caller scope
+      let s1 = "not captured!" ## does not override `s1=2`
+      let xignoredLocal = kfoo4
+      let x3 = newLit kfoo4
+      result = genAst({kNoExposeLocalInjects}, s1=2, s2="asdf", x0=newLit x0, x1=x1, x2, x3):
```

Comment:
s2 is a string, not a literal. This should be a compilation error. Please enforce the usage of ``newLit`` to make clear that only NimNodes can be captured.

Reply:
I fixed it in latest commit (ditto, doesn't appear since you closed PR); `newLit` is now done automatically if needed